### PR TITLE
Initial publication of smp module for Python

### PR DESCRIPTION
**Pull request by anton-malakhov**
@avgorshk

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+from distutils.core import *
+
+setup(  name        ="SMP",
+        description ="Static Multi-Processing",
+        long_description= doc,
+        url         ="https://software.intel.com/intel-distribution-for-python",
+        author      ="Intel Corporation",
+        author_email="scripting@intel.com",
+        license     ="Intel EULA",
```
Comment (fschlimb) [inline setup.py]: shouldn't this be BSD*?

File: smp/__init__.py [hunk 2]
```diff
@@ -0,0 +1,577 @@
...
+import sys
+from threading import Lock
+from multiprocessing.pool import Pool, ThreadPool
+from multiprocessing import cpu_count
+try:
+    from concurrent.futures import ProcessPoolExecutor, ThreadPoolExecutor
+except:
+    ProcessPoolExecutor = None
+    ThreadPoolExecutor = None
+from ctypes import *
```
Comment (fschlimb) [inline smp/__init__.py:39]: uh

File: smp/__init__.py [hunk 3]
```diff
@@ -0,0 +1,577 @@
...
+    cpu_count = len(cpu_list)
+
+    global oversubscription_factor
+    if cpu_count < oversubscription_factor:
+        oversubscription_factor = cpu_count
+
+    threads_per_process = oversubscription_factor
+    if cpu_count >= process_count:
+        threads_per_process = threads_per_process*int(round(float(cpu_count)
+                              /float(process_count)))
```
Comment (fschlimb) [inline smp/__init__.py:229]: I am not sure I understand what this block is really doing.

Generally, the entire file has only 2 comments/hints. I have not been involved in any of the discussions about this topic so it's hard to follow what exactly is happening. The description in the other file does not help much either (as it does not go beyond re-phrasing the names of the arguments). There is no explanation what so ever of what this module is supposed to do, e.g. what the different input arguments really mean and how this works.

I suggest you add a high-level description of this file is really doing and add more specific comment to each function. Things like lines 223-225 could also use some comment.

File: smp/__init__.py [hunk 4]
```diff
@@ -0,0 +1,577 @@
...
+            w.daemon = True
+            w.start()
+            debug('added worker')
+
+def affinity_worker35(inqueue, outqueue, initializer=None, initargs=(),
+                      maxtasks=None, wrap_exception=False,
+                      process_count=1, process_id=0):
+    from multiprocessing.pool import worker
+    set_proc_affinity(process_count, process_id)
+    worker(inqueue, outqueue, initializer, initargs, maxtasks, wrap_exception)
```
Comment (fschlimb) [inline smp/__init__.py:278]: This function is almost identical with the 27 version. I see no reason for this code-duplication.

File: smp/__init__.py [hunk 5]
```diff
@@ -0,0 +1,577 @@
...
+                                   self._initializer,
+                                   self._initargs, self._maxtasksperchild,
+                                   self._wrap_exception,
+                                   self._processes, base_id + i)
+                            )
+            self._pool.append(w)
+            w.name = w.name.replace('Process', 'PoolWorker')
+            w.daemon = True
+            w.start()
+            debug('added worker')
```
Comment (fschlimb) [inline smp/__init__.py:393]: I guess generally all this code duplication should be avoided.

File: smp/__init__.py [hunk 6]
```diff
@@ -0,0 +1,577 @@
...
+    if success:
+        print("done")
+    return 0 if success else 1
+
+def _main():
+    global oversubscription_factor
+    global max_top_workers
+
+    if not sys.platform.startswith('linux'):
+        raise "Only linux is currently supported"
```
Comment (fschlimb) [inline smp/__init__.py:550]: why is that?

**Review by fschlimb — COMMENTED** [review]
The code looks very nice. As noted, I might be able to give feedback on the content once I better understand what it's doing.

File: smp/__init__.py (hunk 2 above)
Comment (anton-malakhov) [reply smp/__init__.py:39]: that's fair.. we'll fix that later though

File: smp/__init__.py (hunk 6 above)
Comment (anton-malakhov) [reply smp/__init__.py:550]: some platform specific is encoded right now, Anton G will rewrite it later

File: smp/__init__.py (hunk 3 above)
Comment (anton-malakhov) [reply smp/__init__.py:229]: @avgorshk
